### ISSUE TITLE
"Update Chrome" dialog

### ISSUE THREAD

**Issue by flokol120** [oja/amazon-music-linux#24]
![image](https://user-images.githubusercontent.com/22056791/50052948-d7124b80-012c-11e9-8ed3-0121859a6e55.png)
Sorry for the german dialog. It says that the used chrome version is deprecated and won't be supported anymore in the near future.
Updating electron might do the trick.

**Comment by flokol120** [comment]
@jcrollard84 I tried to update electron but this didn't solve the issue. With the Beta version of electron I'm getting several errors but the beta version is using a newer chromium version. As soon as I got some time I'll take a deeper look into that.

**Comment by jcrollard84** [comment]
> @jcrollard84 I tried to update electron but this didn't solve the issue. With the Beta version of electron I'm getting several errors but the beta version is using a newer chromium version. As soon as I got some time I'll take a deeper look into that.

Ok, thanks :)

**Comment by blakethepatton** [comment]
Not technically a "fix" rather more of a hack, but you could [override the user agent string](https://stackoverflow.com/a/35672988/3397446) which amazon might be checking to get the version number from. 

```
import { session } from 'electron';

session.defaultSession.webRequest.onBeforeSendHeaders((details, callback) => {
  details.requestHeaders['User-Agent'] = 'SuperDuperAgent';
  callback({ cancel: false, requestHeaders: details.requestHeaders });
});
```

Though if anything I suspect they want to use a feature included in a newer version of chrome. So this fix would maybe only work short term.

**Comment by flokol120** [comment]
Update: The Beta version seems the be using an outdated version as well. Maybe quitting the dialog automatically on start would be a dirty, yet easy workaround.

**Comment by blakethepatton** [comment]
If you're doing a dirty fix, you could try just modifying the user agent. Maybe throw a polyfill in to make sure the latest js features are available...

**Comment by flokol120** [comment]
Update: @blakethepatton unfortunately, modifying the user agent in both the `main.ts` and the `renderer.ts` doesn't seem to do the trick. Maybe they are checking on a new JS feature which aren't implemented yet. I'll try a polyfill as soon as I got some free time.

**Comment by flokol120** [comment]
![image](https://user-images.githubusercontent.com/22056791/51028923-28c7bd80-1595-11e9-82b5-3cf183203bdd.png)
Now there is also stuff going on in the console. I am not entirely sure what is going wrong. Maybe somebody who worked with AWS before can read something out of it.

**Comment by eclyptox** [comment]
Updated to electron 4 and still appearing dialog.

**Comment by slaybhi** [comment]
It stopped showing the close button on the "Update Chrome" box. 
![deepinscreenshot_20190209223646](https://user-images.githubusercontent.com/24721237/52523891-a67a0880-2cbc-11e9-98a9-1e90e04b603a.png)

**Comment by mmilleror** [comment]
I just did a fresh install and received the same message.

**Comment by blakethepatton** [comment]
I'm still seeing the close button. Probably just an incremental rollout.

**Comment by latinomx** [comment]
The same issue 17 feb

**Comment by rovingrambler** [comment]
Ditto. Been getting the warning for a while, today it stopped working.

![image](https://user-images.githubusercontent.com/30183263/53020539-16a72d80-3425-11e9-873e-3bce76be57a1.png)

**Comment by blakethepatton** [comment]
There's a new version of electron out... I tried updating and building but it's my first time messing with electron and when I went to run the app I got nothing. One breaking change that I fixed locally:

main.ts:32
```diff
-const shouldQuit = app.makeSingleInstance((commandLine, workingDirectory) => {
+app.requestSingleInstanceLock()
+const shouldQuit = app.on('second-instance', (event, commandLine, workingDirectory) => {
```

**Comment by flokol120** [comment]
@blakethepatton I stumbled on that too. Though I did not get it working. I even tried to switch to BrowserView but that didn't work either. I found an official amazon music help page where it said specifically that Linux is not supported. Unfortunately I did not found the US version of this, only the german one. I have had the bad feeling that we are specifically blocked out by amazon :disappointed:.

**Comment by blakethepatton** [comment]
Well my issue was that I couldn't launch the electron app after building it... But I did upgrade from `2.0.9` to `4.0.6`. I'm sure that somewhere something changed that prevented me from opening my build... I've just never played with electron before today so I wasn't able to figure out why it wouldn't launch... I'm sure I have a log somewhere, just don't know where to look. I'll probably take a more through look Friday and try again. 

Looks like you need to be running chrome 64+ to get their site working. I doubt it's a direct attack against your app. 

* [Chrome 63](https://www.browserling.com/browse/win/7/chrome/63/https%3A%2F%2Fmusic.amazon.com) 

* [Chrome 64](https://www.browserling.com/browse/win/7/chrome/64/https%3A%2F%2Fmusic.amazon.com)

**Comment by pranavt** [comment]
Blocked from running it today. Does building from the git source allow running the latest version of Electron, and consequently the latest version of the Chrome runtime?

**Comment by blakethepatton** [comment]
@pranavt I tried doing the upgrade of electron locally and building... But the app never launched for me. And I haven't worked with electron enough to know what I did wrong. So to answer your question, no.

**Comment by flokol120** [comment]
> Well my issue was that I couldn't launch the electron app after building it... But I did upgrade from `2.0.9` to `4.0.6`. I'm sure that somewhere something changed that prevented me from opening my build... I've just never played with electron before today so I wasn't able to figure out why it wouldn't launch... I'm sure I have a log somewhere, just don't know where to look. I'll probably take a more through look Friday and try again.
> 
> Looks like you need to be running chrome 64+ to get their site working. I doubt it's a direct attack against your app.
> 
> * [Chrome 63](https://www.browserling.com/browse/win/7/chrome/63/https%3A%2F%2Fmusic.amazon.com)
> * [Chrome 64](https://www.browserling.com/browse/win/7/chrome/64/https%3A%2F%2Fmusic.amazon.com)

If running Chrome was necessary we would be knocked out. I installed the chromium version of the latest electron beta and I was able to listen to music. So Chrome is not required. Nevertheless I could not get it to work through electrons built-in chromium.

**Comment by flokol120** [comment]
Update on this matter:
I took a deeper look into amazon music and found the following:

```
var a = n("HUa0")
          , s = n("ImCZ")
          , o = n("LvDl")
          , l = n("pRuz")
          , c = [{
            dmlsName: "WIDEVINE",
            keySystem: "com.widevine.alpha",
            supportedConfiguration: [{
                initDataTypes: ["cenc"],
                audioCapabilities: [{
                    contentType: 'audio/mp4;codecs="mp4a.40.2"'
                }]
            }]
        }, {
            dmlsName: "PLAYREADY",
            keySystem: "com.microsoft.playready",
            supportedConfiguration: [{
                initDataTypes: ["cenc"],
                audioCapabilities: [{
                    contentType: 'audio/mp4;codecs="mp4a.40.2"'
                }]
            }]
        }]
          , u = (i(r = {}, amznMusic.constants.BROWSER_ENUMS.CHROME, c[0]),
        i(r, amznMusic.constants.BROWSER_ENUMS.FIREFOX, c[0]),
        i(r, amznMusic.constants.BROWSER_ENUMS.OPERA, c[0]),
        i(r, amznMusic.constants.BROWSER_ENUMS.BRAVE, c[0]),
        i(r, amznMusic.constants.BROWSER_ENUMS.IE, c[1]),
        i(r, amznMusic.constants.BROWSER_ENUMS.EDGE, c[1]),
        r);
```
The interesting part is that amazon music seems to be using playready DRM by Microsoft and most importantly Widevine by Google.
Problem with that is electron does not support Widevine on linux. See [electron/#12427](https://github.com/electron/electron/issues/12427#issuecomment-401449104):

> Additionally, the castLabs version will always be behind the electron version and does not support Linux. 

**TL;DR: I am afraid this project seems to be dead.** :disappointed:

**Comment by blakethepatton** [comment]
Thanks for trying! It was a great application while it lasted.

**Comment by rovingrambler** [comment]
Hey just a brief look but I wonder if this election fork is worth looking at.

They're aiming for drop-in replacement.

Seems we're back to the good, bad and ugly days....

On May 1, 2019 2:45 PM, Blake Patton <notifications@github.com> wrote:

Thanks for trying! It was a great application while it lasted.

**Comment by rovingrambler** [comment]
Would help if I pasted the link

https://github.com/castlabs/electron-releases

On May 1, 2019 6:28 PM, hconnolly@gcet.net wrote:
Hey just a brief look but I wonder if this election fork is worth looking at.

They're aiming for drop-in replacement.

Seems we're back to the good, bad and ugly days....

On May 1, 2019 2:45 PM, Blake Patton <notifications@github.com> wrote:

Thanks for trying! It was a great application while it lasted.

**Comment by blakethepatton** [comment]
@rovingrambler 

I think @flokol120 already looked in to the castlabs version. 

> > Additionally, the castLabs version will always be behind the electron version and does not support Linux.

Just looked into why there isn't linux support and allegedly the linux userbase is too small for widevine. 

https://github.com/castlabs/electron-releases/issues/5#issuecomment-403614103
> They believe Linux is not currently a large enough Electron platform to warrant spending the time required to make the required additions to the CDM, so it has been put on ice for now.

**Comment by rovingrambler** [comment]
I've moved to Nuvola Apps.
Working well. Their Amazon desktop app downloads chrome and extracts widevine

https://nuvola.tiliado.eu/
